app/models/curriculum.rb

```diff
@@ -40,28 +40,40 @@ class Curriculum < ApplicationRecord
   # validations
   validates :name, presence: true, uniqueness: { scope: :unit_id }, length: { maximum: 255 }
   validates :programs, presence: true
-  validates :semesters_count, numericality: { greater_than_or_equal_to: 0 }
   validates :status, inclusion: { in: statuses.keys }
+  validate :check_semester
 
   # custom methods
   def build_semesters
-    return false unless valid?
+    return false if semesters_count >= number_of_semesters
```

Comment:
```suggestion
    return if semesters_count >= number_of_semesters
```

Metod ismi `?` ile bitmedigi surece bana `false` donmesini beklemezdim. Bu metodun, adinin belirttigi (`build_semesters`) isi yapip ciktiginda `true`, bir sey yapmadan ciktiginda ise `nil` donmesi daha dogal olur. Hem de metodun return value'sunu kullandigimiz yerlerde `false` kaynakli side-effect'lerle ugrasmak zorunda kalmayiz.

Reply:
@msdundar good catch!

Reply:
Review'den hareketle ekstra bir not düşeyim.  @msdundar'ın

> Metod ismi ? ile bitmedigi surece bana false donmesini beklemezdim.

cümlesi (anlatımda kullanmak için) şöyle de kurulabilir: "bu metod false gibi out-of-band bir değer dönüyor, bunu beklemezdim"

Programlamada sıklıkla karşılaştığımız bir durum, yapılan bir işlemde ortaya çıkan olumsuzlukların (çoğunlukla hatalar ama onunla sınırlı değil) veya nasıl diyelim, farklı durumların çağıran tarafa bildirilmesi.  C gibi "multi value return" veya "exception" özelliklerine sahip olmayan eski ve geleneksel dillerde bu çoğunlukla "out-of-band" bir değer dönerek yapılır.  Örneğin:

- String'de bir karakter arıyor ve bulduğumuzda indisini dönüyoruz, bulamadığımızda "out-of-band" olarak `-1` dönüyoruz.  (Neden "out-of-band"?  Geçerli indis değerleri yani ""band" 0'dan başlayan pozitif sayılar, `-1` geçerli bir indis değeri olamaz, yani "band" dışı)

- String dizisinde (haystack) bir string arıyoruz (needle) ve bulamadığımızda `null` dönüyoruz.  Bu örnekte de `null` "out of band".

Modern yüksek seviyeli dillerde bu pratiklerin yerini çoğunlukla "multi value return" veya "exception", o da olmadı kodun dikkatli şekilde yapılandırmasıyla özel değer dönmekten kaçınmak alıyor.  Genel ilke şu, fonksiyon/metodun olağan dönüş değerini özel şekilde yorumlayarak program akışını değiştirmeye çalışma.  Fonksiyon karnından konuşmamalı, doğrudan ("explicit") olarak durumu bildirebilmeli.

Farklı dönüş durumlarına bakalım:

- Eğer bu bir hata ise "exception" mekanizmalı dillerde önerilen yol hatanın (gerekirse custom bir class yazarak) exception ile bildirilmesi

  ```ruby
  def main
    helal_dönüş_değeri = birşey_yap
    # helal_dönüş_değeri ile başka şeyler yap
  rescue İstisna => e
   # hatayı yönet
  end
  ```
  Multi value dillerde, örneğin Go'da ise:

  ```go
  helal_dönüş_değeri, err = birşey_yap()
  if err != nil {
          // hatayı yönet
  }
  ```

  Kıssadan hisse , hata durumlarını Ruby gibi dillerde "out-of-band" dönüş değerleriyle (çağıran tarafın yorumlasını gerektirecek tarzda) bildirme.  Örneğin aşağıdaki hata yönetimi hatalı.

  ```ruby
  haram_da_olabilecek_dönüş_değeri = birşey_yap
  if haram?  haram_da_olabilecek_dönüş_değeri
    # hatayı yönet
  end
  ```

- Hata değilse ne yapalım?  Örneğin aranan bir şey bulunamadı.  Şöyle mi yapalım?

  ```ruby
  unless (iğne = ara(samanlık)) # veya if (iğne = ara(samanlık)).nil?
    # iğne bulunamadığında (yani nil veya false olduğunda) ne yapılacaksa
  end
  ```

  Örnekteki "iğne" için `nil` (veya duruma göre `false`) bir "out-of-band" değer olduğundan bu kod makul gibi duruyor.  Ama Ruby için konuşacak olursak bu kod yine de "smell" yayıyor.  Şu çok helal fakat:

  ```ruby
  iğneler = ara(samanlık)
  if iğneler.empty?
    # hiç bir iğne bulunamadığında ne yapılacaksa
  end
  ```

  İkinci kod neden kesinlikle helal?  Çünkü `iğneler`'in alabileceği boş değer (`[]`) bir in-band değer.

  Özetle bu soruya her yerde geçerli olabilecek bir cevap veremiyoruz.  Nelerden kaçınılması veya dikkatli düşünülmesi gerektiğini kabaca anlattım (out-of-band değerler).  Duruma göre farklı çözümler uygulanabilir.  Örneğin Ruby'de "multi value return" yok ama "value type"lar var.  Trivial olmayan durumlarda metod bir "value" nesnesi dönebilir (hatayı daima exception ile) ve çağıran taraf da bu value nesnesini sorgulayabilir.  Örneğin ben aşağıdaki paterni çok seviyorum.

  ```ruby
  class Result # value type
    ...
  end

  def karmaşık_iş
    ...
    Result.new x: falan, y: filan, status: :ok, count: count
  end
  ```

Review'e dönecek olursak.  Yukarıda anlatmaya çalıştığım nedenlerden dolayı Ruby'de mantıksal bir değer dönen metodlar hemen hemen daima "predicate" metodlardır (`?` ile sonlandırılan isimler).  Metod predicate değilse (sonda `?` yok) "out-of-band" değer dönme pratiğini kullanan bir "code smell" ihtimalini dikkate alırız.  Her durum hatalıdır diyemem, case case gitmek lazım.

Reply:
>  Örneğin Ruby'de "multi value return" yok ama "value type"lar var. Trivial olmayan durumlarda metod bir "value" nesnesi dönebilir (hatayı daima exception ile) ve çağıran taraf da bu value nesnesini sorgulayabilir. Örneğin ben aşağıdaki paterni çok seviyorum.

Eklemeden geçemeyeceğim.  Dönüş değeri olarak basit bir şey yerine "value type" kullanılması bir başka açıdan bakılırsa [Primitive Obsession](https://refactoring.guru/smells/primitive-obsession)'dan kaçınma oluyor.  Benim sizlere belki de vereceğim en önemli öğüt: "aman ha sakın Primitive Obsession halinde olmayın" olurdu.  Her problem ona yakışan düzgün bir "value type" gerektirir.  Yeri gelir bu bir basit `Struct` veya `Open Struct`'tır, yeri gelir custom bir nesnedir.  Aman ha basit veri tipleriyle (string, hash, array) karmaşık verileri yönetmeye çalışmayın.

Reply:
> Eğer bu bir hata ise "exception" mekanizmalı dillerde önerilen yol hatanın (gerekirse custom bir class yazarak) exception ile bildirilmesi

Buraya ufak bir ekleme, custom class'lar monitorleme ve incident management icin de cok faydali oluyor. Sikca kullanilan bir pattern olarak:

```ruby
class Curriculum < ApplicationRecord
  class MyCustomError < StandardError

  def some_method
    do_something_important
  rescue SomeUnexpectedError => e
    raise MyCustomError, "Mantikli bir hata mesaji ve incelemek icin object #{e.inspect} if @some_condition_satisfied?
  end
end
```